src/dnstest_harness.erl

```diff
@@ -220,16 +351,60 @@ test_header(ExpectedHeader, Response) ->
 % Send the message to the given host:port via UDP.
 send_udp_query(Message, Host, Port) ->
     Packet = dns:encode_message(Message),
-    ?LOG_DEBUG("Sending UDP query to host ~p and port ~p", [host(), port()]),
-    {ok, Socket} = gen_udp:open(0, [binary, {active, false}]),
-    ok = gen_udp:send(Socket, Host, Port, Packet),
-    QueryResponse =
-        case gen_udp:recv(Socket, 65535, 6000) of
-            {ok, {Host, _Port, Reply}} when is_binary(Reply) -> {ok, dns:decode_message(Reply)};
-            {error, Error} -> {error, {Error, {server, {Host, Port}}}}
-        end,
-    gen_udp:close(Socket),
-    QueryResponse.
+    ?LOG_DEBUG("Sending UDP query to host ~p and port ~p", [Host, Port]),
+    case gen_udp:open(0, [binary, {active, false}]) of
+        {ok, Socket} ->
+            try
+                ok = gen_udp:send(Socket, Host, Port, Packet),
+                % Increased buffer size for potential large UDP responses
+                case gen_udp:recv(Socket, 65535, 6000) of
+                    {ok, {_RHost, _RPort, Reply}} when is_binary(Reply) ->
+                        {ok, dns:decode_message(Reply)};
+                    {error, timeout} ->
+                        ?LOG_WARNING("UDP query timed out for ~p", [Message#dns_message.questions]),
+                        {error, {timeout, {server, {Host, Port}}}};
+                    {error, Error} ->
+                        ?LOG_ERROR("UDP recv error: ~p", [Error]),
+                        {error, {Error, {server, {Host, Port}}}}
+                end
+            after
+                gen_udp:close(Socket)
+            end;
+        {error, Reason} ->
+            ?LOG_ERROR("Failed to open UDP socket: ~p", [Reason]),
+            {error, {socket_error, Reason}}
+    end.
+
+% Send the message to the given host:port via TCP.
+send_tcp_query(Message, Host, Port) ->
+    Packet = dns:encode_message(Message),
+    ?LOG_DEBUG("Sending TCP query to host ~p and port ~p", [Host, Port]),
+    case gen_tcp:connect(Host, Port, [binary, {packet, 2}, {active, false}], 1000) of
+        {ok, Socket} ->
+            try
+                ok = gen_tcp:send(Socket, Packet),
+                case gen_tcp:recv(Socket, 0, 6000) of
```

Comment:
Btw, what about this timeout, do we need it that long too? I'd want test to fail fast if something gets stuck 🤔

Reply:
Changed in [fe55edd](https://github.com/dnsimple/dnstest/pull/21/commits/fe55edd45cfdbd6ce8246ab5a7d434206afd0d0e)

Reply:
Why I don't see it as changed? The diff here still tells me 6000 🤔